src/columnstore/columnstore_metadata.cpp

```diff
@@ -297,4 +354,166 @@ string ColumnstoreMetadata::SecretsSearchDeltaOptions(const string &path) {
+        if (txn_id == GetCurrentTransactionId()) {
+            PostgresFunctionGuard(CatalogTupleDelete, table, &tuple->t_self);
+            break;
+        }
+    }
+    systable_endscan_ordered(scan);
+    CommandCounterIncrement();
+    index_close(index, RowExclusiveLock);
+    table_close(table, RowExclusiveLock);
+}
+
+void ColumnstoreMetadata::FlushDeltaRecords(std::function<void(DeltaRecord)> dump_func) {
+    const TimestampTz now = GetCurrentTransactionStartTimestamp();
+    ::Relation table = table_open(DeltaUpdateRecord(), AccessShareLock);
+    ::Relation index = index_open(DeltaUpdateRecordTimestamp(), AccessShareLock);
+
+    // Flush delta records in the order of timestamp.
+    ScanKeyData key[1];
+    ScanKeyInit(&key[0], 2 /*attributeNumber*/, BTGreaterEqualStrategyNumber, F_INT8GT, TimestampGetDatum(0));
+    SysScanDesc scan = systable_beginscan_ordered(table, index, snapshot, 1 /*nkeys*/, key);
+    TupleDesc desc = RelationGetDescr(table);
+    HeapTuple tuple;
+    while (HeapTupleIsValid(tuple = systable_getnext_ordered(scan, ForwardScanDirection))) {
+        bool isnull = false;
+        Datum col;
+
+        col = heap_getattr(tuple, 1 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        const TransactionId txn_id = static_cast<TransactionId>(col);
+
+        // There're two possible states for tuples in delta update record table: ongoing and committed.
+        // Tuples inserted by aborted rows won't exist, because insertion into delta update table happens within
+        // postgres transaction, so pg deletes the tuple at txn aborted.
+        //
+        // If current transaction isn't committed, skip.
+        if (!TransactionIdDidCommit(txn_id)) {
+            continue;
+        }
+
+        // If record timestamp is already in the future, skip.
+        col = heap_getattr(tuple, 2 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        const TimestampTz timestamp = DatumGetTimestampTz(col);
+        if (timestamp >= now) {
+            break;
+        }
+
+        // Only flush committed old delta records.
+        DeltaRecord cur_delta_record;
+        col = heap_getattr(tuple, 4 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        cur_delta_record.path = GetTextFromDatun(col);
+
+        col = heap_getattr(tuple, 5 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        cur_delta_record.delta_options = GetTextFromDatun(col);
+
+        col = heap_getattr(tuple, 6 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        cur_delta_record.file_names = ExtractFileNames(col);
+
+        col = heap_getattr(tuple, 7 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        cur_delta_record.file_sizes = DatumToIntArray<int64_t>(col);
+
+        col = heap_getattr(tuple, 8 /*attnum*/, desc, &isnull);
+        D_ASSERT(!isnull);
+        cur_delta_record.is_add_files = DatumToIntArray<int8_t>(col);
+
+        dump_func(std::move(cur_delta_record));
```

Comment:
Shouldn't we consolidate multiple Postgres transactions into a single one? Currently, it seems that we generate one lake commit for each record, which corresponds to an individual PG transaction.

Reply:
> Currently, it seems that we generate one lake commit for each record

One record could contain multiple file names; in this PR I'm trying to keep the same output as we do now, could  followup in later PRs.